### ISSUE TITLE
`needless_pass_by_value` false positive on non-Copy enum when deconstructing with match

### ISSUE THREAD

**Issue by findepi** [rust-lang/rust-clippy#13320]
### Summary

`needless_pass_by_value` lint is triggered when an enum ?Copy type is moved into a function and consumed by deconstructing with match

### Reproducer

I tried this code:

```rust
// #[derive(Clone, Copy, Debug)]
pub enum MyEnum {
    A,
    B(Option<bool>),
}

pub fn b_as_bool(m: MyEnum) -> bool {
    match m {
        MyEnum::B(Some(b)) => b,
        _ => false,
    }
}
```

i invoke clippy with `cargo clippy -- -A clippy::all -D clippy::needless-pass-by-value`

I saw this happen:

```
error: this argument is passed by value, but not consumed in the function body
 --> src/clippy_needless_pass_by_value.rs:9:21
  |
9 | pub fn b_as_bool(m: MyEnum) -> bool {
  |                     ^^^^^^ help: consider taking a reference instead: `&MyEnum`
  |
help: consider marking this type as `Copy`
 --> src/clippy_needless_pass_by_value.rs:3:1

```

However, if i follow the advise and make `b_as_bool` take `&MyEnum`, the function no longer compiles, so i think clippy advise is not right.

Note:
clippy doesn't flag this problem when `MyEnum` implements `Copy` (not sure why).

### Version

```text
rustc 1.80.1 (3f5fd8dd4 2024-08-06)
binary: rustc
commit-hash: 3f5fd8dd41153bc5fdca9427e9e05be2c767ba23
commit-date: 2024-08-06
host: aarch64-apple-darwin
release: 1.80.1
LLVM version: 18.1.7
```

**Comment by Jarcho** [comment]
Clippy is correct, but has an incomplete suggestion. Deconstructing via a match doesn't move the whole value, only the captured parts. You can do this:

```rust
pub fn b_as_bool(m: &MyEnum) -> bool {
    match *m {
        MyEnum::B(Some(b)) => b,
        _ => false,
    }
}
```

---

`Copy` types aren't linted because they are never moved into a function, so there's no usability issue with taking them by value.

**Comment by findepi** [comment]
@Jarcho thanks you're correct, moving _part_ of the value works, provided that the part is `Copy` (`Option<bool>` is `Copy`, right?)

indeed, if i modify the example replacing `bool` with another `enum` (non-Copy), clippy no longer flags this.


Thanks for the tip!

**Comment by Jarcho** [comment]
`Option<T>` is `Copy` if `T` is `Copy`.
